theano/gof/type.py

```diff
@@ -795,3 +796,241 @@ def signature(self):
+        self.ctype = kwargs.pop('ctype', 'double')
+        self.check_ctype()
+        for k in kwargs:
+            if re.match('^[A-Z][A-Z0-9_]*$', k) is None:
+                raise AttributeError('%s: invalid enum name: "%s". '
+                                     'Only capital letters, underscores and digits '
+                                     'are allowed.' % (type(self).__name__, k))
+            if isinstance(kwargs[k], bool):
+                kwargs[k] = int(kwargs[k])
+            elif not isinstance(kwargs[k], (int, float)):
+                raise ValueError('%s: constant "%s": expected integer or floating value, got "%s".'
+                                 % (type(self).__name__, k, type(kwargs[k]).__name__))
+        super(EnumType, self).__init__(**kwargs)
+
+    def __repr__(self):
+        return '%s(%s)' % (type(self).__name__, ', '.join('%s:%s' % (k, self[k]) for k in sorted(self.keys())))
+
+    def __getattr__(self, key):
+        if key in self:
+            return self[key]
+        if key == 'ctype':
+            return self.ctype
+        return Type.__getattr__(self, key)
+
+    def __setattr__(self, key, value):
+        if key in self:
+            raise NotImplementedError('constant values are immutable.')
+        Type.__setattr__(self, key, value)
+
+    def __setitem__(self, key, value):
+        raise NotImplementedError('constant values are immutable.')
+
+    def __delitem__(self, key):
+        raise NotImplementedError('constant values are immutable.')
+
+    def __hash__(self):
+        # All values are Python basic types, then easy to hash.
+        return hash((type(self), self.ctype) + tuple((k, self[k]) for k in sorted(self.keys())))
+
+    def __eq__(self, other):
+        return (type(self) == type(other) and
+                self.ctype == other.ctype and
+                len(self) == len(other) and
+                all(k in other for k in self) and
+                all(self[k] == other[k] for k in self))
+
+    # EnumType should be used to create constants available in both Python and C code.
+    # However, for convenience, we make sure EnumType can have a value, like other common types,
+    # such that it could be used as-is as an op param.
+    # C type of value is defined in self.ctype.
+
+    def filter(self, data, strict=False, allow_downcast=None):
+        if not strict and isinstance(data, bool):
+            data = int(data)
+        assert isinstance(data, (int, float))
+        return data
+
+    def values_eq(self, a, b):
+        return a == b
+
+    def values_eq_approx(self, a, b):
+        return float(a) == float(b)
+
+    @staticmethod
+    def c_support_macro_code():
+        return """
+        #if PY_MAJOR_VERSION >= 3
+            #ifndef PyInt_Check
+                #define PyInt_Check PyLong_Check
+            #endif
+            #ifndef PyInt_AsLong
+                #define PyInt_AsLong PyLong_AsLong
+            #endif
+        #endif
+        """
```

Comment:
Ok then perhaps you should name it differently so that it doesn't look like a new portion of the C code interface.